### PR TITLE
Add metering for locations

### PR DESCRIPTION
**Pull request by dsainati1**
Works towards https://github.com/dapperlabs/cadence-private-issues/issues/2

This adds metering for `AddressLocation`, `StringLocation`, `ScriptLocation` and `TransactionLocation` values. 
______

- [ ] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [X] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

**Review by turbolent — APPROVED** [review]
Nice!

File: runtime/common/transactionlocation.go [hunk 1]
```diff
@@ -73,13 +78,13 @@ func (l TransactionLocation) MarshalJSON() ([]byte, error) {
...
-		func(typeID string) (location Location, qualifiedIdentifier string, err error) {
-			return decodeTransactionLocationTypeID(typeID)
+		func(gauge MemoryGauge, typeID string) (location Location, qualifiedIdentifier string, err error) {
+			return decodeTransactionLocationTypeID(gauge, typeID)
 		},
 	)
 }
 
-func decodeTransactionLocationTypeID(typeID string) (TransactionLocation, string, error) {
+func decodeTransactionLocationTypeID(gauge MemoryGauge, typeID string) (TransactionLocation, string, error) {
```
Comment (SupunS) [inline runtime/common/transactionlocation.go:87]: Same here, `TransactionLocation` needs to be metered

File: runtime/imported_values_memory_metering_test.go [hunk 2]
```diff
@@ -46,12 +46,13 @@ func TestImportedValueMemoryMetering(t *testing.T) {
...
 
 	runtimeInterface := func(meter map[common.MemoryKind]uint64) *testRuntimeInterface {
-		return &testRuntimeInterface{
+		intf := &testRuntimeInterface{
 			meterMemory: testUseMemory(meter),
-			decodeArgument: func(b []byte, t cadence.Type) (cadence.Value, error) {
-				return jsoncdc.Decode(b)
-			},
 		}
+		intf.decodeArgument = func(b []byte, t cadence.Type) (cadence.Value, error) {
```
Comment (SupunS) [inline runtime/imported_values_memory_metering_test.go:52]: Maybe add some test for metering the locations passed as part arguments (e.g: the ones created via the JSON decoder)

File: runtime/interpreter/decode.go [hunk 3]
```diff
@@ -1724,15 +1724,12 @@ func (d LocationDecoder) decodeAddressLocation() (common.Location, error) {
 		return nil, err
 	}
 
+	common.UseMemory(d.memoryGauge, common.NewBytesMemoryUsage(len(encodedAddress)))
```
Comment (SupunS) [inline runtime/interpreter/decode.go:1727]: `decodeAddressBytes` already does this and the length check. So maybe use that.
Also can remove the TODO above (L#1695)

**Pull request merged**